host/frontend/host-orchestrator/operation.go

```diff
@@ -0,0 +1,143 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"log"
+	"sync"
+)
+
+type NewOperationError string
+
+func (s NewOperationError) Error() string {
+	return fmt.Sprintf("failure creating a new operation: %s", string(s))
+}
+
+type Operation struct {
+	Name   string
+	Done   bool
+	Result OperationResult
+}
+
+type OperationResult struct {
+	Error OperationResultError
+}
+
+type OperationResultError struct {
+	ErrorMsg string
+}
+
+type OperationManager interface {
+	New() Operation
+
+	Get(string) (Operation, bool)
+
+	Complete(string, OperationResult)
+
+	// TODO(b/231319087) This should handle timeout.
+	Wait(string)
+}
+
+type mapOMOperationEntry struct {
+	data  Operation
+	mutex sync.RWMutex
+	wait  sync.WaitGroup
+}
+
+type MapOM struct {
+	uuidFactory func() string
+	mutex       sync.RWMutex
+	operations  map[string]*mapOMOperationEntry
+}
+
+func NewMapOM(uuidFactory func() string) *MapOM {
+	return &MapOM{
+		uuidFactory: uuidFactory,
+		mutex:       sync.RWMutex{},
+		operations:  make(map[string]*mapOMOperationEntry),
+	}
+}
+
+const mapOMNewUUIDRetryLimit = 100
+
+func (m *MapOM) New() Operation {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+	name, retryCount := m.uuidFactory(), 0
+	_, found := m.operations[name]
+	for found && retryCount < mapOMNewUUIDRetryLimit {
+		name, retryCount = m.uuidFactory(), retryCount+1
+		_, found = m.operations[name]
+	}
+	if retryCount == mapOMNewUUIDRetryLimit {
+		panic("uuid retry limit reached")
+	}
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	op := &mapOMOperationEntry{
+		data: Operation{
+			Name:   name,
+			Done:   false,
+			Result: OperationResult{},
+		},
+		mutex: sync.RWMutex{},
+		wait:  wg,
+	}
+	m.operations[name] = op
+	return op.data
+}
+
+func (m *MapOM) Get(name string) (Operation, bool) {
+	op, ok := m.getOperationEntry(name)
+	if !ok {
+		return Operation{}, false
+	}
+	op.mutex.RLock()
```

Comment:
It's unclear what this lock is protecting. The only statement actually covered is the `return` line and is it possible for that to race with anything else?

It's also a bit strange that the lock is effectively acquired twice in most of these methods, first by the inner call to`getOperationEntry` and then later by doing it itself. Is it possible to say "the mutex must be at least read-locked to enter `getOperationEntry`" and then only lock the mutex a single time for these methods?

Reply:
we have the map mutex and different mutexes per each operation entry.

1. the map mutex is for races between New and (Get, Complete), note that Get and Complete do not race against each other as both are read from the map perspective.
2. the operation mutexes are for races between Get and Complete for the same operation.

the two-layers solution is used to avoid blocking Get when Completing each operation.

Reply:
Ah I didn't see that they were distinct mutes, my mistake. I'm still confused about what this particular lock is protecting as I don't see how a return can race with something else.

Reply:
changed the approach of `Named result parameters` but instead used a classic `Lock Read Unlock` approach to improve to have a very clear reading.